google/cloud/storage/transfer_manager.py

```diff
@@ -0,0 +1,188 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import concurrent.futures
+
+import tempfile
+
+
+def upload_many(
+    file_blob_pairs,
+    skip_if_exists=False,
+    upload_kwargs=None,
+    max_workers=None,
+    deadline=None,
+    raise_exception=False,
+):
+    if upload_kwargs is None:
+        upload_kwargs = {}
+    if skip_if_exists:
+        upload_kwargs["if_not_generation_match"] = 0
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+        futures = []
+        for path_or_file, blob in file_blob_pairs:
+            method = (
+                blob.upload_from_filename
+                if isinstance(path_or_file, str)
+                else blob.upload_from_file
+            )
+            futures.append(executor.submit(method, path_or_file, **upload_kwargs))
+    results = []
+    concurrent.futures.wait(
+        futures, timeout=deadline, return_when=concurrent.futures.ALL_COMPLETED
+    )
+    for future in futures:
+        if not raise_exception:
+            exp = future.exception()
+            if exp:
+                results.append(exp)
+                continue
+        results.append(future.result())
+    return results
+
+
+def download_many(
+    blob_file_pairs,
+    download_kwargs=None,
+    max_workers=None,
+    deadline=None,
+    raise_exception=False,
+):
+    if download_kwargs is None:
+        download_kwargs = {}
+    with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+        futures = []
+        for blob, path_or_file in blob_file_pairs:
+            method = (
+                blob.download_to_filename
+                if isinstance(path_or_file, str)
+                else blob.download_to_file
+            )
+            futures.append(executor.submit(method, path_or_file, **download_kwargs))
+    results = []
+    concurrent.futures.wait(
+        futures, timeout=deadline, return_when=concurrent.futures.ALL_COMPLETED
+    )
+    for future in futures:
+        if not raise_exception:
+            exp = future.exception()
+            if exp:
+                results.append(exp)
+                continue
+        results.append(future.result())
+    return results
+
+
+def download_chunks_concurrently_to_file(
+    blob,
+    file_obj,
+    chunk_size=200 * 1024 * 1024,
+    max_workers=None,
+    download_kwargs=None,
+    deadline=None,
+):
+    if download_kwargs is None:
+        download_kwargs = {}
+    # We must know the size of the object, and the generation.
+    if not blob.size or not blob.generation:
+        blob.reload()
+
+    def download_range_via_tempfile(blob, file_obj, start, end, download_kwargs):
+        tmp = tempfile.TemporaryFile()
+        blob.download_to_file(tmp, start=start, end=end, **download_kwargs)
```

Comment:
Default retries are applied for downloads... it seems like retries will behave as intended. Just thinking how if any multiple worker threads might impact retries for this new feature

Reply:
Retries are still configurable using upload/download_kwargs and I expect them to work normally.